methods/matching/find_pairs.py

```diff
@@ -173,8 +177,65 @@ def find_match_iteration(
 
     logging.info("Finished find match iteration")
 
-@jit(nopython=True, fastmath=True, error_model="numpy")
 def make_s_set_mask(
+        rng: np.random.Generator,
+        k_set: pd.DataFrame,
+        m_set: pd.DataFrame,
+        m_dist_thresholded: np.ndarray,
+        k_subset_dist_thresholded: np.ndarray,
+        hard_match_columns: list,
+        required: int,
+        hard_match_categories: dict[Any, np.ndarray]
+    ):
+    s_set_mask_true = np.zeros(m_set.shape[0], dtype=np.bool_)
+    no_potentials = np.zeros(k_set.shape[0], dtype=np.bool_)
+
+    # Split K and M into those categories and create masks
+    for values in hard_match_categories.values():
+        k_selector = np.all(k_set[hard_match_columns] == values, axis=1)
+        m_selector = np.all(m_set[hard_match_columns] == values, axis=1)
+        logging.info("  category: %a |K|: %d |M|: %d", values, k_selector.sum(), m_selector.sum())
+        # Make masks for each of those pairs
+        key_s_set_mask_true, key_no_potentials = make_s_set_mask_rumba_inner(
+            m_dist_thresholded[m_selector],
+            k_subset_dist_thresholded[k_selector],
+            required,
+            rng
+        )
+        # Merge into one s_set_mask_true
+        s_set_mask_true[m_selector] = key_s_set_mask_true
+        # Merge into no_potentials
+        no_potentials[k_selector] = key_no_potentials
+    return s_set_mask_true,no_potentials
+
+def make_s_set_mask_rumba_inner(
+    m_dist_thresholded: np.ndarray,
+    k_set_dist_thresholded: np.ndarray,
+    required: int,
+    rng: np.random.Generator
+):
+    k_size = k_set_dist_thresholded.shape[0]
+    m_size = m_dist_thresholded.shape[0]
+
+    s_include = np.zeros(m_size, dtype=np.bool_)
+    k_miss = np.zeros(k_size, dtype=np.bool_)
+
+    m_tree = make_kdrangetree(m_dist_thresholded, np.ones(m_dist_thresholded.shape[1]))
+
+    rumba_tree = make_rumba_tree(m_tree, m_dist_thresholded)
+
+    for k in range(k_size):
+        k_row =  k_set_dist_thresholded[k]
+        possible_s = rumba_tree.members_sample(k_row, required, rng)
```

Comment:
This as I read it is likely to be 10 items, but not guaranteed, and it's also unlikely but not impossible that we get the same values for each k, which means that s_include will have less than the target number of items in it (this same may have been true of the old code). I Guess we should add some warning or fail if that's the case? I'm not sure at what point the size of S not being 10xK is an issue.

Reply:
That is the case with the old code. I don't know if there is a level we should warn or fail at? But especially for the smaller hard match categories we definitely can't guarantee any matching points.